src/components/code/code_block_annotations.tsx

```
@@ -0,0 +1,85 @@
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import React, { FunctionComponent, ReactNode, useState } from 'react';
+
+import { useEuiTheme } from '../../services';
+import { useEuiI18n } from '../i18n';
+import { EuiPopover } from '../popover';
+import { EuiIcon } from '../icon';
+import { useEuiButtonFocusCSS } from '../../themes/amsterdam/global_styling/mixins/button';
+
+import { euiCodeBlockAnnotationsStyles } from './code_block_annotations.style';
+
+export type LineAnnotationMap = Record<number, ReactNode>;
+
+export const EuiCodeBlockAnnotation: FunctionComponent<{
+  lineNumber: number;
+}> = ({ lineNumber, children }) => {
+  const [isOpen, setIsOpen] = useState(false);
+
+  const ariaLabel = useEuiI18n(
+    'euiCodeBlockAnnotations.ariaLabel',
+    'Click to view a code annotation for line {lineNumber}',
+    { lineNumber }
+  );
+
+  const { euiTheme, colorMode } = useEuiTheme();
+  const styles = euiCodeBlockAnnotationsStyles(euiTheme);
+  const buttonIconFocusStyle = useEuiButtonFocusCSS();
+  const cssButtonIconStyles = [
+    styles.euiCodeBlockAnnotation__buttonIcon,
+    buttonIconFocusStyle,
+  ];
+
+  const isDarkMode = colorMode === 'DARK';
+
+  return (
+    <EuiPopover
+      isOpen={isOpen}
+      closePopover={() => setIsOpen(false)}
+      button={
+        <button
+          onClick={() => setIsOpen(!isOpen)}
+          aria-label={ariaLabel}
+          css={cssButtonIconStyles}
+          data-test-subj="euiCodeBlockAnnotationIcon"
+        >
+          <EuiIcon
+            type={AnnotationInfoIcon}
+            size="s"
+            color={isDarkMode ? euiTheme.colors.ink : 'ghost'}
+          />
+        </button>
+      }
+      css={styles.euiCodeBlockAnnotation}
+      zIndex={Number(euiTheme.levels.mask) + 1} // Ensure fullscreen annotation popovers sit above the mask
+      anchorPosition="leftCenter"
```

Thank you for this component!

Literally only comment is a nit:

We'd prefer this to be `downLeft`. There isn't necessarily a lot of space to the left of code blocks and then it defaults to `rightCenter` and covers up the code snippet it is annotating. I guess you could make the same argument for multi-line annotations but I think the more common case is one line so we'd rather optimize for that.

---

I'm totally good with that! Do you want to make a super quick PR with that change? 🙌

---

Will do!

---

https://github.com/elastic/eui/pull/6600 🎉